### PR TITLE
Update github actions

### LINKED ISSUE #6: Update GH actions
Creating this issue more as a note / reminder - realized when I was reviewing #5. The `pre-commit` was updated to be more aligned with the template (in #4), but had missed updating the actions to do the same.

Feel free to assign this to me @clane9.

### PR DESCRIPTION
**Pull request by kaitj**
Updates actions to be more in line with the template repository, but uses a matrix for unit testing all of the supported python versions. I haven't tested `codecov` but I'd presume that should work as long as its been enabled in the repo settings. Closes #6.

`mypy` was throwing a bunch of warnings and errors, so I made a few changes to the codebase to address these. I thought we had already addressed these in #4, but maybe it was just the `pre-commit` catching the changed files. Would be good if someone else can run the `mypy` command on the `main` branch just to confirm that it actually is throwing an error. Was mostly type redefinitions and some mixed usage of types for `coord` (`np.ndarray` vs `tuple` vs `Coord`).

### CODE REVIEW

**Comment by nx10** [comment]
I've sometimes noticed a disparity between pre-commit mypy and system mypy too. Maybe pre-commit does not read pyptoject.toml 🤔

**Review by clane9 — CHANGES_REQUESTED** [review]
Thanks @kaitj! Thanks for making all the type fixes. Some of mypy's complaints are kind of pedantic and I don't love the result. But I guess it is still good to have as we try to advance the project. Otherwise, I left a couple minor comments regarding the CI workflow.

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup python environment
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: pip
+      - name: Install dependencies
+        run: |
+          python -m venv py${{ matrix.python-version }}-venv
```
[the remark this answers is not in the document]
Comment (kaitj) [reply .github/workflows/test.yaml:24]: Not necessary at all, we can do this without a virtual environment.

File: src/niclips/image/_pad.py [hunk 2]
```diff
@@ -13,7 +14,7 @@ class Align(Enum):
 
 
 def pad_to_size(
-    img: np.ndarray,
+    img: np.ndarray | Image.Image,
```
[the remark this answers is not in the document]
Comment (kaitj) [reply src/niclips/image/_pad.py:17]: I think the styles changed with py3.10 ([PEP484](https://peps.python.org/pep-0484) vs [PEP604](https://peps.python.org/pep-0604/). Do you have a preference for one or the other?

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          source py${{ matrix.python-version }}-venv/bin/activate
+          pip install -U pip
+          pip install -e .\[dev,test\]
```
[the remark this answers is not in the document]
Comment (kaitj) [reply .github/workflows/test.yaml:27]: No preference for me here. My thought was to try to catch any issues early on, but if we are also testing on a specific version, I think it probably makes more sense to just use the `requirements.txt`

**Comment by kaitj** [comment]
> I've sometimes noticed a disparity between pre-commit mypy and system mypy too. Maybe pre-commit does not read pyptoject.toml 🤔

Hmm, wonder if this would still be the case if pre-commit installs the local / system mypy instead of the remote repository.

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    branches:
+      - main
+  pull_request:
+
+jobs:
+  unit:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
[the remark this answers is not in the document]
Comment (kaitj) [reply .github/workflows/test.yaml]: Hmm, I may create a separate branch to test the actions so as to not spam others inboxes (do other people get the failure emails too?) 

`bids2table` can't be found when using the `requirements.txt`

File: .github/workflows/test.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (clane9) [reply .github/workflows/test.yaml]: Ok, I agree with you. Let's keep as is.

File: src/niclips/image/_pad.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clane9) [reply src/niclips/image/_pad.py:17]: My sense is trying to support old versions of python is getting annoying. Do you see a downside to moving the minimal version to 3.10 and at the same time upgrade the type annotation style?

[the remark this answers is not in the document]
Comment (kaitj) [reply src/niclips/image/_pad.py:17]: Nothing that jumps out at me.

[the remark this answers is not in the document]
Comment (clane9) [reply src/niclips/image/_pad.py:17]: Ok let's do that then.

File: .github/workflows/test.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaitj) [reply .github/workflows/test.yaml]: Sounds good - I did change it to remove the matrix (re: one of the comments you left below) and to just use the python version specified in the `pyproject.toml`, but I don't anticipate any major issues not testing every single python version (especially if we bring up minimum python version).

File: src/niclips/typing.py [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
-from typing import Tuple, Union
+from typing import Union
 
 import nibabel as nib
 import numpy as np
 
 StrPath = Union[str, os.PathLike]
 
-Coord = Tuple[float, float, float]
+Coord = Union[tuple[float, float, float] | np.ndarray]
```
Comment (clane9) [inline src/niclips/typing.py:11]: This looks odd, is it correct?

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -14,7 +14,7 @@ repos:
   rev: v1.9.0
   hooks:
   - id: mypy
-    args: [--ignore-missing-imports]
+    args: [--ignore-missing-imports, --allow-redefinition]
```
Comment (clane9) [inline .pre-commit-config.yaml:17]: Cool, I did not know this was an option.

File: src/niclips/figures/multi_view.py [hunk 7]
```diff
@@ -1,6 +1,8 @@
 """Generation of different multi-views."""
 
-from typing import List, Optional, Tuple
+from __future__ import annotations
```
Comment (clane9) [inline src/niclips/figures/multi_view.py:3]: Is this necessary still?

File: src/niclips/image/_pad.py [hunk 8]
```diff
@@ -29,23 +29,25 @@ def pad_to_size(
     padding = size - current_size
     align = Align(align)
     if align in {Align.LEFT, Align.TOP}:
-        padding = (0, padding)
-    elif align == {Align.RIGHT, Align.BOTTOM}:
-        padding = (padding, 0)
+        padding: tuple[int, int] = (0, padding)
```
Comment (clane9) [inline src/niclips/image/_pad.py:32]: Why is this annotation needed? I guess bc padding is being redefined?

**Review by clane9 — APPROVED** [review]
Ok last few comments then good to go. Thank you so much!

File: src/niclips/typing.py (hunk 5 above)
Comment (kaitj) [reply src/niclips/typing.py:11]: Hmm, this is one place where it doesn't look like the formatter caught the updates. I'll go through these.

**Comment by kaitj** [comment]
Okay, I think this one is finally good to go. I left the `_version.py` file untouched, which is the only one that still uses the old typehinting style.

Strange to me that both the `ruff` formatter and `pyupgrade` didn't catch the issues with the imports. Will have to look into that more.

**Comment by nx10** [comment]
Why not `importlib.metadata.version('niftyone')`?

**Pull request merged**